File: server/src/grpc/song.rs
use api::pb::song_service_server::SongService;
use api::pb::{
    CreateSongRequest, DeleteSongRequest, GetSongRequest, ListSongsRequest, ListSongsResponse,
    Song, UpdateSongRequest,
};
use tonic::{Request, Response, Result, Status};

#[derive(Debug, Default)]
pub struct SongServer;

#[tonic::async_trait]
impl SongService for SongServer {
    async fn create_song(
        &self,
        _request: Request<CreateSongRequest>,
    ) -> Result<Response<Song>, Status> {
        todo!()
    }

    async fn get_song(&self, _request: Request<GetSongRequest>) -> Result<Response<Song>, Status> {
        todo!()
    }

    async fn list_songs(
        &self,
        _request: Request<ListSongsRequest>,
    ) -> Result<Response<ListSongsResponse>, Status> {
        todo!()
    }

    async fn update_song(
        &self,
        _request: Request<UpdateSongRequest>,
    ) -> Result<Response<Song>, Status> {
        todo!()
    }

    async fn delete_song(
        &self,
        _request: Request<DeleteSongRequest>,
    ) -> Result<Response<()>, Status> {
        todo!()
    }
}

File: backend/internal/api/song.go
package api

import (
	"context"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	songpb "musicclubbot/backend/proto/song"
	emptypb "google.golang.org/protobuf/types/known/emptypb"
)

// SongService implements song catalog endpoints.
type SongService struct {
	songpb.UnimplementedSongServiceServer
}

func (s *SongService) ListSongs(ctx context.Context, req *songpb.ListSongsRequest) (*songpb.ListSongsResponse, error) {
	return nil, status.Errorf(codes.Unimplemented, "ListSongs not yet implemented")
}

func (s *SongService) GetSong(ctx context.Context, req *songpb.SongId) (*songpb.SongDetails, error) {
	return nil, status.Errorf(codes.Unimplemented, "GetSong not yet implemented")
}

func (s *SongService) CreateSong(ctx context.Context, req *songpb.CreateSongRequest) (*songpb.SongDetails, error) {
	return nil, status.Errorf(codes.Unimplemented, "CreateSong not yet implemented")
}

func (s *SongService) UpdateSong(ctx context.Context, req *songpb.UpdateSongRequest) (*songpb.SongDetails, error) {
	return nil, status.Errorf(codes.Unimplemented, "UpdateSong not yet implemented")
}

func (s *SongService) DeleteSong(ctx context.Context, req *songpb.SongId) (*emptypb.Empty, error) {
	return nil, status.Errorf(codes.Unimplemented, "DeleteSong not yet implemented")
}

func (s *SongService) JoinRole(ctx context.Context, req *songpb.JoinRoleRequest) (*songpb.SongDetails, error) {
	return nil, status.Errorf(codes.Unimplemented, "JoinRole not yet implemented")
}

func (s *SongService) LeaveRole(ctx context.Context, req *songpb.LeaveRoleRequest) (*songpb.SongDetails, error) {
	return nil, status.Errorf(codes.Unimplemented, "LeaveRole not yet implemented")
}
